### PR TITLE
elb_target_group: Add support for protocol_version parameter

### PR DESCRIPTION
**Pull request by sharvarikhedkar**
##### SUMMARY
Added support for `protocol_version` param in `elb_target_group`.

Fixes [1422](https://github.com/ansible-collections/community.aws/issues/1422).

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME
`elb_target_group`

### CODE REVIEW

File: plugins/modules/elb_target_group.py [hunk 1]
```diff
@@ -566,6 +581,7 @@ def create_or_update_target_group(connection, module):
     params['TargetType'] = target_type
     if target_type != "lambda":
         params['Protocol'] = module.params.get("protocol").upper()
+        params['ProtocolVersion'] = module.params.get('protocol_version')
```
Comment (goneri) [inline plugins/modules/elb_target_group.py:584]: Hi @sharvarikhedkar,

Thank you for your PR.

From the CI:

```
2022-09-19 02:29:45.787246 | controller |   File "/tmp/ansible_elb_target_group_payload_gni90h2b/ansible_elb_target_group_payload.zip/ansible_collections/community/aws/plugins/modules/elb_target_group.py", line 822, in create_or_update_target_group
2022-09-19 02:29:45.787254 | controller |   File "/tmp/ansible_elb_target_group_payload_gni90h2b/ansible_elb_target_group_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/modules.py", line 327, in deciding_wrapper
2022-09-19 02:29:45.787260 | controller |     return retrying_wrapper(*args, **kwargs)
2022-09-19 02:29:45.787267 | controller |   File "/tmp/ansible_elb_target_group_payload_gni90h2b/ansible_elb_target_group_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/cloud.py", line 119, in _retry_wrapper
2022-09-19 02:29:45.787274 | controller |     return _retry_func(
2022-09-19 02:29:45.787280 | controller |   File "/tmp/ansible_elb_target_group_payload_gni90h2b/ansible_elb_target_group_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/cloud.py", line 69, in _retry_func
2022-09-19 02:29:45.787287 | controller |     return func()
2022-09-19 02:29:45.787294 | controller |   File "/home/zuul/venv/lib64/python3.9/site-packages/botocore/client.py", line 386, in _api_call
2022-09-19 02:29:45.787300 | controller |     return self._make_api_call(operation_name, kwargs)
2022-09-19 02:29:45.787307 | controller |   File "/home/zuul/venv/lib64/python3.9/site-packages/botocore/client.py", line 677, in _make_api_call
2022-09-19 02:29:45.787313 | controller |     request_dict = self._convert_to_request_dict(
2022-09-19 02:29:45.787320 | controller |   File "/home/zuul/venv/lib64/python3.9/site-packages/botocore/client.py", line 725, in _convert_to_request_dict
2022-09-19 02:29:45.787326 | controller |     request_dict = self._serializer.serialize_to_request(
2022-09-19 02:29:45.787334 | controller |   File "/home/zuul/venv/lib64/python3.9/site-packages/botocore/validate.py", line 319, in serialize_to_request
2022-09-19 02:29:45.787341 | controller |     raise ParamValidationError(report=report.generate_report())
2022-09-19 02:29:45.787347 | controller | botocore.exceptions.ParamValidationError: Parameter validation failed:
2022-09-19 02:29:45.787354 | controller | Invalid type for parameter ProtocolVersion, value: None, type: <class 'NoneType'>, valid types: <class 'str'>
```

`ProtocolVersion` cannot be `None`. It has to be a string. So I would suggest to set the value if it's `!= None`.

https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/elbv2.html#ElasticLoadBalancingv2.Client.create_target_group

Also, you will need a changelog fragment for a change like this.

**Review by goneri — CHANGES_REQUESTED** [review]

File: plugins/modules/elb_target_group.py (hunk 1 above)
Comment (sharvarikhedkar) [reply plugins/modules/elb_target_group.py]: Hi @goneri, thank you for the review. 
I've added a condition to read the `protocol_version` param only when provided and also added a changelog fragment. Thanks.

Comment (markuman) [reply plugins/modules/elb_target_group.py]: outdated, I resolve this conversation

File: plugins/modules/elb_target_group.py [hunk 2]
```diff
@@ -89,6 +89,12 @@
     required: false
     choices: [ 'http', 'https', 'tcp', 'tls', 'udp', 'tcp_udp', 'HTTP', 'HTTPS', 'TCP', 'TLS', 'UDP', 'TCP_UDP']
     type: str
+  protocol_version:
+    description:
+      - Specifies protocol version.
+    required: false
+    choices: ['GRPC', 'HTTP1', 'HTTP2']
+    type: str
```
Comment (markuman) [inline plugins/modules/elb_target_group.py:97]:
```suggestion
  protocol_version:
    description:
      - Specifies protocol version.
    required: false
    choices: ['GRPC', 'HTTP1', 'HTTP2']
    type: str
    version_added: 4.3.0
```

But I'm not sure if we are fast enough here ...maybe it becomes 5.1.0

**Comment by markuman** [comment]
...`backport-4` label is valid until `5.0.0` is released. Then `backport-4` label will be removed wird `backport-5` label.

@sharvarikhedkar thank you for the PR.  
Can you also expand the integrationtest here https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/elb_target/tasks/alb_target.yml and test the parameter?

File: changelogs/fragments/1496-elb_target_group-add-protocol_version-support.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - elb_target_group - add support for ProtocolVersion param (https://github.com/ansible-collections/community.aws/pull/1496).
```
Comment (markuman) [inline changelogs/fragments/1496-elb_target_group-add-protocol_version-support.yml:2]:
```suggestion
minor_changes:
  - elb_target_group - add support for ``protocol_version`` parameter (https://github.com/ansible-collections/community.aws/pull/1496).
```

Comment (sharvarikhedkar) [reply changelogs/fragments/1496-elb_target_group-add-protocol_version-support.yml]: Fixed.

File: plugins/modules/elb_target_group.py (hunk 2 above)
Comment (sharvarikhedkar) [reply plugins/modules/elb_target_group.py:97]: Hi @markuman, thank you for the suggestion. I've added `4.3.0` for now, if `4.3.0` gets released before this PR is merged, I'll modify it to `5.1.0`.

Comment (markuman) [reply plugins/modules/elb_target_group.py:97]: Yes, it must be now `5.1.0` @sharvarikhedkar

File: plugins/modules/elb_target_group.py [hunk 4]
```diff
@@ -89,6 +89,13 @@
...
+    version_added: 4.3.0
```
Comment (markuman) [inline plugins/modules/elb_target_group.py:98]:
```suggestion
    version_added: 5.1.0
```

File: plugins/modules/elb_target_group.py (hunk 2 above)
Comment (sharvarikhedkar) [reply plugins/modules/elb_target_group.py:97]: Thank you @markuman, updated it to `5.1.0`

**Comment by sharvarikhedkar** [comment]
> Can you also expand the integrationtest here https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/elb_target/tasks/alb_target.yml and test the parameter?

@markuman added tests here https://github.com/ansible-collections/community.aws/pull/1496/files#diff-585e159c50673676a30c7c81dfb4c4428b7ad174b9183f8d7bc77f518f70f302.

File: tests/integration/targets/elb_target/tasks/etg_protocol_version.yml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        state: present
+      register: create_result
+
+    - assert:
+        that:
+          - create_result is not changed
+          - create_result is not failed
+          - create_result.protocol_version == "GRPC"
+          - create_result.protocol == "HTTP"
+          - create_result.port == 80
```
Comment (markuman) [inline tests/integration/targets/elb_target/tasks/etg_protocol_version.yml:51]: @sharvarikhedkar can you also please a third test that verify that a change of the `protocol_version` works properly?

Comment (markuman) [reply tests/integration/targets/elb_target/tasks/etg_protocol_version.yml:51]: ...otherwise LGTM

**Review by goneri — APPROVED** [review]

File: tests/integration/targets/elb_target/tasks/etg_protocol_version.yml (hunk 5 above)
Comment (sharvarikhedkar) [reply tests/integration/targets/elb_target/tasks/etg_protocol_version.yml:51]: Hi @markuman, I tried testing the code for [updating the target group with different protocol_versions](https://github.com/ansible-collections/community.aws/pull/1496/commits/54b2128980451f7f7d07a84ec303d832a0e67f58) but [looking at the boto3 documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/elbv2.html#ElasticLoadBalancingv2.Client.modify_target_group) and trying to modify `protocol_version` through AWS console, it seems like the update operation for `protocol_version` parameter is not supported yet, any insights on this?

Comment (markuman) [reply tests/integration/targets/elb_target/tasks/etg_protocol_version.yml:51]: Oh, I didn't know that. Maybe that should be mentioned than in the module documentation @sharvarikhedkar

Comment (sharvarikhedkar) [reply tests/integration/targets/elb_target/tasks/etg_protocol_version.yml:51]: Added.

**Comment by markuman** [comment]
recheck

**Review by markuman — APPROVED** [review]

**Pull request merged**